K8s multiline logs (#926)
logzio/logz-docs: _source/_includes/log-shipping/multiline-logs-filebeat.md

## Multiline logs

If your original logs span multiple lines, you may find that they arrive in your Logz.io account split into several partial logs. You have the option to use either an explicit configuration or an annotation to configure Filebeat to concatenate multiline logs and avoid having them separated into multiple log documents.

> You will need to provide a regex expression that can identify the first log in a multiline log. Filebeat will concatenate the log lines that follow until it identifies the next log that matches the regex expression. In other words, there is no explicit regex expression to demarcate the _end_ of a multiline log.
> 

Reviewer: > Filebeat will concatenate the log lines that follow until it identifies the next log that matches the regex expression.

That's not accurate, for the example this will be the behaviour, but it's because we set it to work that way, it's not the default behaviour.

```suggestion
You will need to provide a regex expression that can identify the first log in a multiline log. Filebeat will concatenate the log lines based on your configuration settings. In other words, there is no explicit regex expression to demarcate the _end_ of a multiline log.

```
or something like that?

This behavior is managed differently depending on your deployment method:

* If you are using autodiscover hints & annotations, add an annotation to your deployment.
* If you are not using autodiscover, use an explicit configuration.

> 

Reviewer: There's a major benefit of using the hints system here, and I think we should mentions it, so it maybe will help them better to consider whether they want to use filebeat with or without autodiscover:
if they use hints, they can manage the regexes for each component separately, not having to create a regex that will match all the possible cases that they need, like they would need to do when using the regular config without autodiscover.
For example, if i have one pod that I know will need regex X, and another that will need regex Y- in the regular config i'll have to create a regex that can "catch" both X and Y, and if I have another pod with regex Z, i'll need to reconfigure filebeat. But if i'm using hints & annotations, i'll just have to add to pod X annotation with regex X, to pod Y annotation with regex Y, etc.. without needing to change anything on Filebeat itself.

Author: addressed.

### Using an explicit configuration to concatenate multiline logs

> 

Reviewer: Please add somewhere in this section a reference to Filebeat's page that has further and needed explanations on the options those configs have
https://www.elastic.co/guide/en/beats/filebeat/7.8/multiline-examples.html

Open your `filebeat.yml` configuration file in a text editor. Under `filebeat.input`, add a multiline configuration to identify the first line in a multiline log. Filebeat will concatenate all subsequent lines until the `multiline.pattern` identifies the next match. For example:

> 

Reviewer: Same here like in the first comment, it's right for this specific example, but not generally, so maybe it's worth a rephrase or to make it clear that it behaves like this because we configured it to like this, but they can also configure it to behave in different ways, depending on their needs.

  processors:
    - add_kubernetes_metadata:
        host: ${NODE_NAME}
        matchers:
        - logs_path:
            logs_path: "/var/log/containers/"
  multiline.type: pattern
  multiline.pattern: '^[0-9]{4}-[0-9]{2}-[0-9]{2}'
  multiline.negate: true
  multiline.match: after
```

### Using hints & annotations to concatenate multiline logs

If you're using Filebeat autodiscover hints, you can use annotations to identify multiline logs and concatenate them.

You will need to first configure Filebeat to enable the hints system, and add annotations to the relevant components when you deploy them to your cluster.

<div class="tasklist">

##### Enable Filebeat's hints system

First, enable Filebeat's hints system. In your `filebeat.yml` file, set `hints.enabled: true` under the `filebeat.autodiscover` section. For example:

```shell
filebeat.autodiscover:
	providers:
	  - type: kubernetes
	    node: ${NODE_NAME}
	    hints.enabled: true # This part enables the hints
	    hints.default_config:
	      type: container
	      paths:
	        - /var/log/containers/*-${data.kubernetes.container.id}.log
```

##### Add multiline annotations to your deployment

> 

Reviewer: Please add a reference to Filebeat's page about annotations, since it has more explanations about what each annotation does and the options they have.
https://www.elastic.co/guide/en/beats/filebeat/7.8/configuration-autodiscover-hints.html
